src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD114AvoidReturningNullTaskAnalyzer.cs

```diff
@@ -33,28 +33,37 @@ public override void Initialize(AnalysisContext context)
             context.EnableConcurrentExecution();
             context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
 
-            context.RegisterOperationBlockStartAction(Utils.DebuggableWrapper(context => AnalyzeOperationBlockStart(context)));
+            context.RegisterOperationAction(Utils.DebuggableWrapper(context => AnalyzerReturnOperation(context)), OperationKind.Return);
         }
 
-        private static void AnalyzeOperationBlockStart(OperationBlockStartAnalysisContext context)
+        private static void AnalyzerReturnOperation(OperationAnalysisContext context)
         {
-            if (context.OwningSymbol is IMethodSymbol method &&
+            var returnOperation = (IReturnOperation)context.Operation;
+
+            if (returnOperation.ReturnedValue is { ConstantValue: { HasValue: true, Value: null } } && // could be null for implicit returns
+                returnOperation.ReturnedValue.Syntax is { } returnedValueSyntax &&
+                Utils.GetContainingFunction(returnOperation) is { } block &&
+                FindOwningSymbol(block, context.ContainingSymbol) is { } method &&
                 !method.IsAsync &&
                 Utils.IsTask(method.ReturnType))
             {
-                context.RegisterOperationAction(Utils.DebuggableWrapper(context => AnalyzerReturnOperation(context)), OperationKind.Return);
+                context.ReportDiagnostic(Diagnostic.Create(Descriptor, returnedValueSyntax.GetLocation()));
             }
         }
 
-        private static void AnalyzerReturnOperation(OperationAnalysisContext context)
+        private static IMethodSymbol? FindOwningSymbol(IBlockOperation block, ISymbol containingSymbol)
         {
-            var returnOperation = (IReturnOperation)context.Operation;
-
-            if (returnOperation.ReturnedValue is { ConstantValue: { HasValue: true, Value: null } } && // could be null for implicit returns
-                returnOperation.ReturnedValue.Syntax is { } returnedValueSyntax)
+            return block.Parent switch
             {
-                context.ReportDiagnostic(Diagnostic.Create(Descriptor, returnedValueSyntax.GetLocation()));
-            }
+                ILocalFunctionOperation localFunction => localFunction.Symbol,
+                IAnonymousFunctionOperation anonymousFunction => anonymousFunction.Symbol,
+
+                // Block parent is the method declaration, for vbnet this means a null parent but for C# it's a IMethodBodyOperation
```

Comment:
@sharwell, @AArnott Do you think that's worth creating a ticket on roslyn? It would be simpler to have a common behavior for both languages when using operations.

Reply:
🤔 I don't know. I'll leave that up to @sharwell.